examples/opensora_hpcai/opensora/acceleration/communications.py

```diff
@@ -0,0 +1,92 @@
+from typing import Callable, Tuple
+
+import mindspore.nn as nn
+import mindspore.ops as ops
+from mindspore import Tensor
+from mindspore.communication import GlobalComm, get_group_size, get_rank
+
+try:
+    from typing import Literal
+except ImportError:
+    from typing_extensions import Literal
+
+
+def _split(x: Tensor, dim: int, rank: int, world_size: int) -> Tensor:
+    dim_size = x.shape[dim]
+    tensor_list = x.split(dim_size // world_size, axis=dim)
+    x = tensor_list[rank]
+    return x
+
+
+def _communicate_along_dim(x: Tensor, dim: int, func: Callable[[Tensor], Tensor]) -> Tensor:
+    x = x.swapaxes(0, dim)
+    x = func(x)
+    x = x.swapaxes(dim, 0)
+    return x
+
+
+class SplitFowardGatherBackward(nn.Cell):
+    def __init__(
+        self, dim: int = 0, grad_scale: Literal["up", "down"] = "down", group: str = GlobalComm.WORLD_COMM_GROUP
+    ) -> None:
+        super().__init__()
+        self.dim = dim
+        self.rank = get_rank(group)
+        self.world_size = get_group_size(group)
+        self.gather = ops.AllGather(group=group)
+
+        if grad_scale == "up":
+            self.scale = self.world_size
+        else:
+            self.scale = 1 / self.world_size
+
+    def construct(self, x: Tensor) -> Tensor:
+        return _split(x, self.dim, self.rank, self.world_size)
+
+    def bprop(self, x: Tensor, out: Tensor, dout: Tensor) -> Tuple[Tensor]:
+        dout = dout * self.scale
+        dout = _communicate_along_dim(dout, self.dim, self.gather)
+        return (dout,)
+
+
+class GatherFowardSplitBackward(nn.Cell):
+    def __init__(
+        self, dim: int = 0, grad_scale: Literal["up", "down"] = "up", group: str = GlobalComm.WORLD_COMM_GROUP
+    ) -> None:
+        super().__init__()
+        self.dim = dim
+        self.rank = get_rank(group)
+        self.world_size = get_group_size(group)
+        self.gather = ops.AllGather(group=group)
+
+        if grad_scale == "up":
+            self.scale = self.world_size
+        else:
+            self.scale = 1 / self.world_size
+
+    def construct(self, x: Tensor) -> Tensor:
+        x = _communicate_along_dim(x, self.dim, self.gather)
+        return x
+
+    def bprop(self, x: Tensor, out: Tensor, dout: Tensor) -> Tuple[Tensor]:
+        dout = dout * self.scale
+        dout = _split(dout, self.dim, self.rank, self.world_size)
+        return (dout,)
```

Comment:
Is an explicit `bprop` necessary? I recall that Ring Attention requires a more complicated backprop, but even in that case, MindSpore's autograd can handle it.

Reply:
for split and gather, the `bprop` is necessary, otherwise the gradient is not correct based on the unit test; for all2all, actually it is not necessary, but as a easier understanding and align with official code, it is explicitly written

Reply:
removed all2all backward since now it has extra functionality of padding and unpadding, add backward will not messy and uncesscary